### ISSUE TITLE
Provide informative messages when encountering orphaned Components

### ISSUE THREAD

**Issue by aseth1** [opensim-org/opensim-core#1209]
Although the use of `auto` is convenient it is leading to more errors in user code. When a method returns a typed pointer `auto` yields a correctly typed pointer, however, if the type returns a `const T&` (reference) auto is happy to give you a copy of type T if you do not explicitly request a reference, `auto&`.  This is very problematic for all our methods that return references into a Model with a context (order, topology, connections) since a copy is technically not a part of that Model. The copy is "orphaned" (has no parent) but otherwise acts like any other Component until you try to hook it up to anything. First, calls to its finalizeFromProperties, connect, ... will not be invoked automatically from its parent (it has none). If connect is called directly on an orphan its dependencies cannot be found (it isn't part of anything) OR things dependent on it cannot find it OR, worst of all, the original component is found (satisfying the dependency) but all the property changes exist(ed) on an orphaned component that has no effect on your Model. This can happen easily in range for loops (e.g. `for(auto musc : model.updComponentList<Muscle>) { ...` where you think you are editing each muscle in the Model when you are in fact editing a copy of each muscle which will be out of scope at the end of the loop.  

This issue first came up a couple of times at the developer's workshop and recently appeared in a litany of Collin's recent problems. "Who knew one little '&' could wreak so much havoc" he said. No kidding.

I don't think this issue can be easily prevented (short of using the PIMPL idiom) but it would be possible to determine if you were calling connect() on an orphan (easiest) and flag dependencies that cannot be found as potential orphans (but could be for other reasons like typos). But because we routinely edit properties before we add a Component to its parent, we cannot prevent the editing of accidentally orphaned components, which can be the most annoying/difficult to debug.

If you have a potential solution please raise it.

**Comment by tkuchida** [comment]
Some ideas come to mind… don't know whether any of these would be appropriate or feasible. Tagged with :apple:, :banana:, and :cherries: for commenting.

## Strict stages to model building
Could be implemented locally or globally.
  1. :apple: *Local* – Each Component has knowledge of whether or not it is an orphan, and its properties/connectors cannot be modified until it has been adopted by another Component (Device, Model, World, …). Alternatively, the Component could have knowledge of whether it had ever been adopted and an exception could be thrown from the destructor when deleting a Component that was always an orphan, which could indicate that the Component went out of scope unexpectedly (though perhaps there are situations where it is acceptable to delete a Component that was never adopted).
  2. :banana: *Global* – All Components must be added to the (single) "master" object (World, top-level Model, …) before properties/connectors can be modified; properties/connectors of orphaned Components cannot be modified. The end of the "construction" stage (`addComponent()` calls) could be signaled with a call like `World::lockComponentLists()`, or could be called automatically when a property/connector is modified. An exception would be thrown if the user attempts to add a Component when the lock is engaged. Ensuring that all Components are registered with (added to) a "master" object would also fit with the philosophy of an omniscient root object ([switchboard operator](https://pbs.twimg.com/media/CmNGM2UWYAAmrq3.jpg)) that makes all the calls to `connect()`.

## Components are automatically registered with World
:cherries: This design follows from the philosophy of UUIDs (see #1192). On construction, a Component is registered with the (single) "master" object (World, top-level Model, …). The user must make an explicit call to "release" the Component from World before the Component goes out of scope and gets deleted. The Component's destructor would throw an exception (and display an instructive message) when attempting to delete a Component that is still registered with World.

**Comment by chrisdembia** [comment]
We could encourage the use of `auto&&`, which I think may solve some issues. I don't fully understand `auto&&` yet, though. http://stackoverflow.com/questions/13230480/what-does-auto-tell-us

**Comment by aymanhab** [comment]
My first thought (after the shock) is to abandon using auto in user facing code/examples. This would help users who copy a block of code to modify, and also makes the C++ code parallels the scripting code. Will also help avoid narly classes appearing across the interface.

If the auto is using an operator to return the copy, can we "delete" that operator?

**Comment by chrisdembia** [comment]
> If the auto is using an operator to return the copy, can we "delete" that operator?

There are also good reasons to use the copy constructor.

**Comment by aseth1** [comment]
> My first thought (after the shock) is to abandon using auto in user facing code/examples. This would help users who copy a block of code to modify, and also makes the C++ code parallels the scripting code. Will also help avoid narly classes appearing across the interface.

:+1:  It would be helpful for those reasons and worth doing IMHO. It doesn't solve the issue (could forget `&` on fully qualified type) but empirically that wasn't such a big issue before we started using `auto`, so hopefully it would reduce misuse.

**Comment by jenhicks** [comment]
Partly addressed with #1511
